[Adam Ho] iP

adam-ky opened the pull request:
### Duke Chatbot
**What can it do?**
1. Manage your to-do list
2. Keep track of deadlines
3. Record events you'll be attending

**What *exactly* can it do?**
- [X] Add a task
- [X] Delete a task
- [X] Show you your list of task
- [ ] Clear all your tasks (coming soon)

**Why should I use it?**
- Intuitive user experience 💯 
- Delegate task management 📋  
- Simply fun 😆 

> Keeping track of my everyday todos has never been more fun!

Still not convinced? Click [here](https://github.com/adam-ky/ip/) to find out more!

Interested to know more as a developer? Here's a peek at how Duke adds a task via the `execute` method 😮 
```java
public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
    try {
        tasks.addTask(task);
        ui.showAddTask(tasks, task);
        storage.saveTasks(tasks);
    } catch (IOException e) {
        throw new DukeException("Oops!!! somthing went wrong :/");
    }
}
```

Commit:
Add Gradle support

adam-ky committed:
Added greet, echo, and exit functionalities.

adam-ky committed:
Added tasks list and list functionality.

adam-ky committed:
Added Task class and mark tasks as done functionality

adam-ky committed:
Added 3 types of Task - Todo, Deadline, Event

adam-ky committed:
Created TaskManager class to manage tasks

adam-ky committed:
Semi-automate tested Duke

adam-ky committed:
Added DukeException class to handle exceptions

adam-ky committed:
Added delete functionality

adam-ky committed:
Added Enum

adam-ky committed:
Added javadoc comments

adam-ky committed:
Create DukeTest class for unit testing

adam-ky committed:
Add TaskListWriter and TaskListReader classes to save and load data

adam-ky committed:
Add TaskListWriter and TaskListReader

adam-ky committed:
Update naming

adam-ky committed:
Create duke.txt

adam-ky committed:
Add save and load data functionalities

The saveTaskList method iterates through the arraylist of tasks and formats them into a string representation of the tasks in duke.txt.
The loadTaskList method scans through duke.txt and adds the tasks back into the arraylist of tasks during rerun of Duke.

adam-ky committed:
Add the ability to format dates

The formatDate method converts a date from "yyyy-mm-dd" format to "MMM d yyyy" format.

adam-ky committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/TaskManager.java

adam-ky committed:
Add JavaDocs

adam-ky committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/TaskManager.java

adam-ky committed:
Add 3 new classes: Storage, TaskList, and Ui

The Ui class deals with interactions with the user.
The Storage class deals with loading tasks from the file and saving tasks in the file.
The TaskList class contains the task list e.g., it has operations to add/delete tasks in the list

adam-ky committed:
Organize classes into package

adam-ky committed:
Modify Parser and add ListCommand class

Parser class now works for commands: list, todo, deadline, event only.

adam-ky committed:
Add DeleteCommand class

adam-ky committed:
Rectify Exception Handling

adam-ky committed:
Divide classes into packages

Classes are further divided into duke.task and duke.command packages.

adam-ky committed:
Add DoneCommand

DoneCommand class handles marking a task as complete.

adam-ky committed:
Add ExitCommand class

adam-ky committed:
Add Junit Tests

adam-ky committed:
Package the App as a JAR file

adam-ky committed:
Add JavaDocs and fix missing user task description

A user may only provide the command word to chatbot but fail to provide task description. In that case, an Duke Exception is thrown.

adam-ky committed:
Modify coding standards

adam-ky committed:
Create FindCommand to handle finding tasks by keyword

The FindCommand class checks if the task description contains the keyword.
Currently, it does not support finding tasks through the date and time description of a task (i.e. by field of Deadline and at field of Event)

adam-ky committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java

adam-ky committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	data/duke.txt

adam-ky committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Leofeng10 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,62 @@
+
+    public void showAddTask(TaskList tasks, Task task) {
+        System.out.println("    Got it. I've added this task: \n      " + task);
+        System.out.println("    Now you have " + tasks.getListSize() + " in the list.");
```
Should the space in the string be separated from the text?

Leofeng10 commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,35 @@
+ * @author Adam Ho
+ */
+public class Event extends Task {
+    protected String at;
```
is "at" a good variable name for a string that representing a time?

Leofeng10 reviewed (commented):
Great work done. Nice code structure. Very readable code! 👍 Actually, I can't really find something that break coding standards.

adam-ky committed:
Modify exception handling for Storage saveTask method

A try-catch block is added within saveTask method to handle IOException, and try-catch blocks in Command sub-classes have been removed.

crypto-code commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+     * manage a to-do task list. The chatbot stops running when the
+     * user inputs the exit command.
+     */
+    public void run() {
```
The code is well written and clean. It's well divided and packaged to make it easy to use.

crypto-code commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,43 @@
+
+public class Parser {
+
+    public static Command parse(String fullCommand) throws DukeException {
```
Good use of an abstract Command Class to handle all commands.

crypto-code commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,78 @@
+            }
+            return taskList;
+        } catch (FileNotFoundException e) {
+            throw new DukeException("Oops something went wrong :(");
```
You could try to make the error a bit more specific so as to help the user identify the problem.

crypto-code reviewed (commented):
Good Job Overall 👍 Couldn't find any coding standard violations 😄

adam-ky committed:
Add config files

adam-ky committed:
Fix Checkstyle configuration and error handling

adam-ky committed:
Merge branch 'checkstyle-issue'

adam-ky committed:
Fix coding standards

Minor correction in coding standards in accordance to Checkstyle.

adam-ky committed:
Merge branch 'fix-coding-standards'

adam-ky committed:
Add clear task list functionality

A new ClearCommand class has been added to manage whether the user wants to clear his/her list of tasks.
A confirmation message is provided before the task list is actually cleared. If the user says "yes" or "y", then the
task list is cleared. Otherwise, the original task list remains.

adam-ky committed:
Add Gradle to automate project builds

adam-ky committed:
Add JavaFX

adam-ky committed:
Merge branch 'A-Gradle'

adam-ky committed:
Optimize import statements

adam-ky committed:
Merge branch 'branch-A-CheckStyle'

adam-ky committed:
Add GUI to Duke

adam-ky committed:
Merge branch 'branch-Level-10'

adam-ky committed:
Add Java varargs

adam-ky committed:
Merge branch 'branch-A-Varargs'

adam-ky committed:
Add assertions

FindCommand class does not check for cases where the task list containing keywords might be longer than the original task list.

A keyword task list that is longer than the original task list suggests a flaw in the logic or even possibly an infinitely long list is being created.

An assert expression is added before returning the keyword task list.

This helps to ensure that, in the event of an infinite loop, the program stops if keyword task list begins to grow beyond the original task list.

adam-ky committed:
Merge pull request #1 from adam-ky/branch-A-Assertions

Add assertions

adam-ky committed:
Improve Code Quality

Some of code quality changes as per guideline:
* Avoid complicated expression
* Avoid magic numbers
* Avoid long methods

adam-ky committed:
Add Streams feature

Stream is used to filter out the task list when provided with a single
keyword.

adam-ky committed:
Merge pull request #3 from adam-ky/branch-A-CodeQuality

Improve Code Quality

adam-ky committed:
Merge branch 'branch-A-Streams'

# Conflicts:
#	data/duke.txt
#	src/main/java/duke/command/FindCommand.java

adam-ky committed:
Set up CI

adam-ky committed:
Update gradle.yml

adam-ky committed:
Add JavaDoc comments

adam-ky committed:
Finalise the features

adam-ky committed:
Update README.md

adam-ky committed:
Finalise features

adam-ky committed:
Merge branch 'master' of https://github.com/adam-ky/ip

adam-ky committed:
Update README.md

adam-ky committed:
Modify FindCommand logic

The FindCommand's execute checks for the length of keywords list.

The updated logic does not require this check.

adam-ky committed:
Modify message constants

Magic literals within Ui methods are refactored into static constants.

adam-ky committed:
Improve GUI

Tweak padding, background colour, image view shape, app icon and
paddings.

adam-ky committed:
Add SortCommand

The user is able to type "sort" to execute the command. However, the
logic for SortCommand#execution in this version is not implemented
correctly yet.

adam-ky committed:
Add newline at end of file

adam-ky committed:
Rename README.md to doc/README.md

adam-ky committed:
Add date formatting

Users are now required to input date format as dd/mm/yyyy when
specifying date for Deadline and Event tasks.

adam-ky committed:
Replace date type to LocalDate

Instead of a String, Event's at date is now represented by LocalDate
type

adam-ky committed:
Update README.md

adam-ky committed:
Merge remote-tracking branch 'origin/master'

adam-ky committed:
Merge branch 'master' into branch-C-Sort

# Conflicts:
#	data/taskdata.txt
#	src/main/java/captain/task/TaskList.java

adam-ky committed:
Remove TaskList#sort

adam-ky committed:
Add AddCommandParser

Previously, the Parser class alone handles all sort of input from user
to return the corresponding Command.

The Parser class has been renamed to TaskParser and only handles which
CommandParser to return based on the user input command. The
corresponding CommandParser will handle the remaining parse logic
to return the correct Command.

adam-ky committed:
Change product screenshot

adam-ky committed:
Merge branch 'refactor-parser'

adam-ky committed:
Add AddCommand

AddCommand class adds a task into the task list

adam-ky committed:
Update README.md

Instead of three separate commands to add three different types of tasks, the user now only needs to depend on the add command keyword.
